src/libstd/collections/hash/map.rs

```diff
@@ -550,17 +572,25 @@ impl<K, V, S> HashMap<K, V, S>
         where K: Borrow<Q>,
               Q: Eq + Hash
     {
-        let hash = self.make_hash(q);
-        search_hashed(&self.table, hash, |k| q.eq(k.borrow()))
+        if self.table.capacity() != 0 {
```

Comment:
Shouldn't this be checking for size() for the optimization to apply for any empty map? My understanding is that checking for 0 capacity only works for not yet allocated maps.

Reply:
This seems to be the explanation for that: https://github.com/rust-lang/rust/issues/38880#issuecomment-361676600 

As it is written, it's a rearrangement of code that should result in no extra branches.

If I understand correctly, `self.table.capacity() != 0` is a precondition. We can simplify this by using `.size()` instead of it (with a comment), and still have equivalent code — but not in the path that we use to find a VacantEntry in an empty map. It would be good to use the `.size() != 0` check where possible. Here's it's also important to keep the cross-method logical dependencies clear.

Reply:
> Shouldn't this be checking for size() for the optimization to apply for any empty map?

This is what I wanted to do, but as @bluss points out, the problem is that when the map is empty but allocated (as opposed to unallocated) and it doesn't find a key, it is supposed to return `InternalEntry::Vacant`. The crucial problem here is that `InternalEntry::Vacant` requires the hash. This means that you can't get away from computing the hash when the `HashMap` is allocated, because otherwise you can't return an `InternalEntry::Vacant`.

This is also a problem for linear search as suggested in #38880. (I did not fully realise this problem until after writing this pull request.) Since an `InternalEntry::Vacant` requires the hash value, a linear search cannot avoid computing the hash value, so it is not possible to achieve a speedup by avoiding the computation of a hash value. I think this only way to fix this is to change the semantics of `InternalEntry::Vacant`. (Which would require a not-insignificant reworking, but would allow for the `size() == 0` that @arthurprs is considering.)

Reply:
N.B. I did actually try the `size() == 0` aka `is_empty()` check first, before I realised about the `InternalEntry::Vacant` semantics: [caaabe](https://github.com/technicalguy/rust/commit/caaabe5ca89db7d494a6399567d1b4489307bea2)

Reply:
Makes sense. HashMaps are hard 😄

Reply:
Regarding the linear search, for that to yield a net win the backing table would need to support real fast iteration (packed KVs), otherwise iterating the buckets gets more expensive than the hashing. Potentially of interest for https://github.com/bluss/ordermap

Reply:
I did some benchmarks for a linear search, and I think it might still be advantageous for some cases even if the map is not densely packed. Here's my reasoning – tell me what you think:

Let *m* be the allocated size of the hashmap and let *n* be the number of items it contains. Let *S* be the cost of accessing the next bucket in the hashmap (regardless of if said bucket is empty), let *E* be the cost of calculating equality for two elements of a given type *__T__*, and let *H* be the cost of calculating the hash for an element of a given type *__T__*. Assume that *E* and *H* are constant across all elements of type *__T__* (this is not true for all types, e.g. strings).

If *mS* + *nE* < *H* then it is faster to perform a linear search lookup for some key than to compute the hash and perform a direct lookup. The average case of a linear search for a key that exists in the map is half of the worst case.

Now if you have a simple type (e.g. integers) then equality is very cheap and constant time, and computing the hash is (surprisingly) expensive (perhaps because of the DoS resilience?), so a linear search is faster for integer-keyed hashmaps with size <= ~ 20 items.

It gets more complicated for other types, such as strings, because the equality operation doesn't take constant time and/or the hash operation doesn't take constant time. (Although for strings I think it is still the case that the hash operation always takes longer than the equality operation.)

Of course all of this is moot if `InternalEntry::Vacant` requires returning the hash value.

Reply:
Maybe, it depends on a lot of variables, specially an expensive hasher.

You can try something this, but I'm afraid calculating VacantEntryState on misses is too expensive.

```
if self.table.capacity() < CC {
    if table.capacity() == 0:
        return InternalEntry::TableIsEmpty;
    } else {
        // search_linear returns Option<InternalEntry::Occupied>
        search_linear(&self.table, |k| q.eq(k.borrow())).unwrap_or_else(|| {
            self.make_hash(q);
            InternalEntry::Vacant { hash, VacantEntryState::...{...} }
         })
} else {
        let hash = self.make_hash(q);
        search_hashed_nonempty(&mut self.table, hash, |k| q.eq(k.borrow()))
}
```

Reply:
So then it would be faster when searching for keys that exist, but slower when searching for keys that don't exist, right? I don't think that optimisation is worth it if it slows down the case when a key is not found.